### PR TITLE
feat: Add a REST API Client package

### LINKED ISSUE #509: Add a REST API Client package
Create an API Client as a monorepo package to connect to some public API.

- Qiita
- https://openweathermap.org/
- GitHub
- pub.dev

### PR DESCRIPTION
**Pull request by k-nkmr**
## 🔗 Related Issues

- closes #509 

## 🙌 What's Done

- Add a REST API Client package

## ✍️ What's Not Done

- None

## 🖼️ Image Differences

- Display a list of data retrieved from the API.
- Load and display additional data when scrolling to the bottom of the list.
- Perform a keyword search and display the results in a list.
- Show detailed information when an item is selected.

https://github.com/user-attachments/assets/0d0fb8d9-b0d0-4c44-bff7-34a29951388c

## 🤼 Desired Review Method

- [ ] Correction Commit
- [ ] Pair programming

> [!NOTE]
> It is possible that a reviewer's will may cause a method to be implemented that is not selected.

## Pre-launch Checklist
- [x] I have reviewed my own code.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I updated/added relevant documentation (doc comments with ///).

### CODE REVIEW

**Comment by riscait** [comment]
@k-nkmr 
It is easier to review if you have a rough idea of the overall structure!
For example, directory structure, etc. ✍️

**Review by riscait — CHANGES_REQUESTED** [review]
I don't think you'll be adding multiple REST API Clients to `rest_api_client`,
so I think it's best to use a package name that represents the contents,
like `PubDevApiClient`👍

**Comment by k-nkmr** [comment]
```
packages
  ├flutter_app
	└ features
		└pub_dev_packages ⭐️New
			├query: Add providers to fetch packages via PubDevApiClient
			└view: Add a page to displays packages
  ├pub_dev_api_client ⭐️New
	├models: Add models for API responses
	└test
```

@riscait 
>It is easier to review if you have a rough idea of the overall structure!
For example, directory structure, etc. ✍️

I’ve roughly represented the directory structure as a tree diagram, with some parts omitted.
If anything is unclear, feel free to let me know, and I’ll provide additional explanations.🙇🏻‍♀️

>I don't think you'll be adding multiple REST API Clients to rest_api_client,
so I think it's best to use a package name that represents the contents,
like PubDevApiClient👍

I have taken your suggestion and renamed the package to PubDevApiClient🙏🏻

File: packages/flutter_app/lib/features/pub_dev_packages/query/pub_dev_packages_page_state_provider.dart [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import 'package:pub_dev_api_client/pub_dev_api_client.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+import '../../../package_adaptor/pub_dev_api_client_provider.dart';
+
+part 'pub_dev_packages_page_state_provider.g.dart';
+
+@Riverpod(keepAlive: true)
+class PackageSearchWordState extends _$PackageSearchWordState {
```
Comment (riscait) [inline packages/flutter_app/lib/features/pub_dev_packages/query/pub_dev_packages_page_state_provider.dart:9]: badge
I would like Doc comments for Public member👌

Would it be better to use `pubDevPackage` here too?

File: packages/flutter_app/lib/features/pub_dev_packages/view/pub_dev_packages_page.dart [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+  const PubDevPackagesPage({super.key});
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final t = Translations.of(context);
+    final asyncValue = ref.watch(pubDevPackagesPageStateProvider);
+
+    return GestureDetector(
+      behavior: HitTestBehavior.opaque,
+      onTap: FocusScope.of(context).unfocus,
```
Comment (riscait) [inline packages/flutter_app/lib/features/pub_dev_packages/view/pub_dev_packages_page.dart:22]: badge

The `UnfocusOnTap` widget could be used👍

File: packages/pub_dev_api_client/pubspec.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+dependencies:
+  dio: ^5.7.0
+  flutter:
+    sdk: flutter
+  freezed_annotation: ^2.4.4
+  mock_web_server: ^5.0.0-nullsafety.1
+  mockito: ^5.4.4
+  retrofit: ^4.4.1
+  url_launcher: ^6.3.1
```
Comment (riscait) [inline packages/pub_dev_api_client/pubspec.yaml:16]: badge
Maybe `pub_dev_api_client` doesn't use `url_launcher`?

**Review by riscait — CHANGES_REQUESTED** [review]
Sorry I'm late! I left some comments🙏

File: packages/flutter_app/lib/features/pub_dev_packages/query/package_details_provider.dart [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import 'package:pub_dev_api_client/pub_dev_api_client.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+import '../../../package_adaptor/pub_dev_api_client_provider.dart';
+
+part 'package_details_provider.g.dart';
+
+/// Provider for package details by package name.
+@riverpod
+Future<GetPackageDetailsResponseBody> packageDetails(
```
[the remark this answers is not in the document]
Comment (k-nkmr) [reply packages/flutter_app/lib/features/pub_dev_packages/query/package_details_provider.dart]: I’ve incorporated the naming change to use pubDevPackageDetailsProvider, as it makes the purpose much clearer.

File: packages/flutter_app/lib/features/pub_dev_packages/view/pub_dev_packages_page.dart (hunk 2 above)
Comment (k-nkmr) [reply packages/flutter_app/lib/features/pub_dev_packages/view/pub_dev_packages_page.dart]: I’ve updated the code to use the UnfocusOnTap widget.

**Comment by k-nkmr** [comment]
@riscait 

I’ve added additional commits, so I’d appreciate it if you could review them again🙇🏻‍♀️

File: packages/flutter_app/lib/features/pub_dev_packages/query/package_details_provider.dart
Comment (riscait) [inline packages/flutter_app/lib/features/pub_dev_packages/query/package_details_provider.dart:1]: badge
Please change the file name as well👍

**Review by riscait — APPROVED** [review]
LGTM

Thanks!

**Pull request merged**